### PR TITLE
Animation PR

### PR DESCRIPTION
**Pull request by fdev31**
Borrowed most of the code from FTLabs which was the only one I found without graphical glitches.

- Fixed some obvious merge problems in the HEAD of this fork
- Adapted the code to support animations in a more generic way (desktop change is still a bit heuristic)

Some of the code couldn't be made clean, reusing the current APIs, with some hints I can fix it.

It supports:

- open/close (by window type)
- special transient hint support for popup windows
- slide windows on desktop change
- resize/move window
- support "special" desktops to be ignored (for instance: Qtile's scratchpads)

This could eventually cover #635 and #217

Most of the code is from FTLabs' fork, I mostly changed some structures & the win.c file.

### CODE REVIEW

**Comment by absolutelynothelix** [comment]
is @FT-Labs ok with «borrowing» his work? i thought he’ll eventually consider his fork stable and make a pull request himself.

**Comment by fdev31** [comment]
Hum,
I didn't consider that given what I saw in the code... with references to dwm and comments such as "animations will not work out of dwm". For instance for me this fork doesn't display transient windows...
Also I didn't consider it being "borrowing", since I don't claim or hide anything and it's opensource... but I made this PR maybe to revive a discussion on the best options.
I'm a bit shy with communication, but you are right, we should involve them at least for awareness.
I didn't see any contact on the github page, is there a way to "poke" someone in a comment ? Else maybe I can open a ticket to open this topic on the FTlabs fork side ?

**Comment by absolutelynothelix** [comment]
i’ve already poked him with a mention. let’s see what he’ll say and if he’s ok with this, i’ll start reviewing. for now you can (actually must, if you want this to be merged) fix all ci’s failures (until ci’s failing for unknown reasons).

**Comment by fdev31** [comment]
Thank you!
I'll wait for some feedback eventually, if it's a dead end I'll not fix the CI... and as you'll see there is a little more to fix in my code before it can really be a candidate for merging ;) For instance I struggled defining the atoms in picom style and had to do direct xcb calls...

**Comment by absolutelynothelix** [comment]
you may want to draft this pull request if there’s still work to be done.

**Comment by fdev31** [comment]
So it depends, technically it's ready for review... but I know the review will not pass at least for items I wrote, but for those remaining I need some assistance to make it clean, it works, but not everything with the general style.

I already cleaned my code, the remaining bits are mistakes or lack of competence/knowledge in this programming type.

I didn't even know draft pull requests were added actually, what is the advantage ? (I can update my commits as well in a normal PR isn't it ?)

**Comment by absolutelynothelix** [comment]
ok, let’s postpone any fixes while waiting for @FT-Labs reply to prevent wasting of at least your time.

draft pull requests basically indicate that there’s still work to be done and they can’t be merged. i usually make my pull requests draft when they’re technically ready for review or merging, but i still want/need to fix something. it’s optional, though.

**Comment by r0-zero** [comment]
Hey guys just saw this.

I am open for anyone wants to help, but there is some work to do.

i) Shadows needs to be fixed
ii) You seem to be using old code, please update your fork. It uses xrandr now.
iii) dwm related places needs to be changed

I'm working on it too, just too busy these days. Once these problems are solved, it will be good to go.

**Comment by fdev31** [comment]
Thank you for the quick feedback & amazing work!

1. I didn't notice any issue with shadows and probably don't have the skills to fix anything there unless I spend a lot of time
2. I used https://github.com/FT-Labs/picom/commit/1c7466c07ce2e1d6e134fb76548249774dc02ac4 as a base which seems to be 3 weeks old (next branch), the "nextt" branch looked older. Did I miss something ?
3. I changed dwm related places - but had to drop the "minimize" animation since I couldn't find a generic way to do it

I had time & motivation for a week, I can't commit for the future but I'ld like to see better animations on picom and willing to help.

**Comment by fdev31** [comment]
It also fails on the same test (299) on my computer, with or without my PR, on similar check but even earlier... (line 77).

**Comment by fdev31** [comment]
I pushed a style fix and the test 299 is passing now (but still not locally).

**Comment by absolutelynothelix** [comment]
fyi, to fix the style quickly you’re supposed to use `clang-format -i /path/to/file`. i usually run it on all files that i change before comitting.

**Comment by fdev31** [comment]
I had the same idea but unfortunately I have to undo many chunks which are not supposed to be in the PR... so in the end it turned to be more time consuming. Adding such rules late makes things a bit more tricky...

**Comment by shasherazi** [comment]
bump?

**Comment by r0-zero** [comment]
It would be great if someone can check this with different window managers! Thanks

https://github.com/yshui/picom/issues/1052

**Comment by yshui** [comment]
also cc @dccsillag and @Arian8j2

File: src/config.h [hunk 1]
```diff
@@ -40,6 +40,24 @@ enum backend {
 	NUM_BKEND,
 };
 
+enum open_window_animation {
```
Comment (Arian8j2) [inline src/config.h:43]: imo `window_animation` is more suited name

File: src/options.c [hunk 2]
```diff
@@ -181,6 +181,17 @@ static const struct picom_option picom_options[] = {
...
+    {"animation-stiffness-tag-change", required_argument, 806, NULL        , "Animation speed when tag changes (change to a new desktop)."},
+    {"animation-dampening"         , required_argument, 807, NULL          , "Animation dampening ratio (spring dampening as an example)."},
+    {"animation-window-mass"       , required_argument, 808, NULL          , "Animation window mass (lower mass makes animations bumpy)."},
+    {"animation-clamping"          , no_argument      , 809, NULL          , "Enable/disable animation clamping. Disabling increases performance"},
+    {"animation-for-open-window"   , required_argument, 810, NULL          , "Set animation for opening window (Check sample.conf for options)."},
+    {"animation-for-transient-window", required_argument, 811, NULL        , "Set animation for transient (child) windows."},
+    // blacklist ??
+    {"animation-for-tag-change", required_argument    , 813, NULL          , "Set animation for switching desktops."},
+    {"animation-extra-desktops", required_argument    , 814, NULL          , "N desktops will not be considered as standard desktops (useful for some window managers)."},
```
Comment (Arian8j2) [inline src/options.c:194]: no command for `--animation-for-unmap-window`?

File: src/win.c [hunk 3]
```diff
@@ -461,6 +478,203 @@ static void win_update_properties(session_t *ps, struct managed_win *w) {
...
+	} else {
+		animation = ps->o.wintype_option[w->window_type].animation;
+	}
+
+	/* START OF ANIMATION CODE (make it optional) */
+
+	if (animation == OPEN_WINDOW_ANIMATION_INVALID) {
+		if (transient_window) {
+			int16_t mx, my;
+			animation = OPEN_WINDOW_ANIMATION_SLIDE_DOWN; //ps->o.animation_for_transient_window;
```
Comment (Arian8j2) [inline src/win.c:515]: opps :smile:

File: src/win.c [hunk 4]
```diff
@@ -461,6 +478,203 @@ static void win_update_properties(session_t *ps, struct managed_win *w) {
...
+			if (get_mouse_position(ps, &mx, &my)) {
+				if (my >= (randr_mon_center_y*2) - w->pending_g.height) {
+					animation = OPEN_WINDOW_ANIMATION_SLIDE_UP;
```
Comment (Arian8j2) [inline src/win.c:518]: here too

File: src/config.h [hunk 5]
```diff
@@ -172,6 +192,32 @@ typedef struct options {
...
+	/// Spring stiffness for current tag animation
+	double animation_stiffness_tag_change;
+	/// Window mass for animation
+	double animation_window_mass;
+	/// Animation dampening
+	double animation_dampening;
+	/// Whether to clamp animations
+	bool animation_clamping;
+	/// Animation blacklist. A linked list of conditions.
+	c2_lptr_t *animation_blacklist;
```
Comment (Arian8j2) [inline src/config.h:219]: `animation_exclude` is more suitable

File: src/options.c [hunk 6]
```diff
@@ -181,6 +181,17 @@ static const struct picom_option picom_options[] = {
...
                                                                              "window is fullscreen based only on its size and coordinates."},
+    {"animations"                  ,no_argument       , 804, NULL          , "Enable/disable animations."},
+    {"animation-stiffness-in-tag"  , required_argument, 805, NULL          , "Animation speed in current tag (float)."},
+    {"animation-stiffness-tag-change", required_argument, 806, NULL        , "Animation speed when tag changes (change to a new desktop)."},
+    {"animation-dampening"         , required_argument, 807, NULL          , "Animation dampening ratio (spring dampening as an example)."},
+    {"animation-window-mass"       , required_argument, 808, NULL          , "Animation window mass (lower mass makes animations bumpy)."},
+    {"animation-clamping"          , no_argument      , 809, NULL          , "Enable/disable animation clamping. Disabling increases performance"},
+    {"animation-for-open-window"   , required_argument, 810, NULL          , "Set animation for opening window (Check sample.conf for options)."},
```
Comment (Arian8j2) [inline src/options.c:190]: where is `sample.conf` :smile:

**Review by Arian8j2 — COMMENTED** [review]
really good animation system, good job, things i think is missing:
- add commands and descriptions to `man` docs
- add examples to `picom.sample.conf`
- animation rule (like opacity rule), eg: can declare specific open animation for apps that match a rule

also it has some bugs with my setup with `dwm` and multiple monitors, windows in second monitor will come all the way from first monitor to second, tried `animation-extra-desktops` but it didn't help it

File: src/options.c (hunk 2 above)
Comment (fdev31) [reply src/options.c:194]: Indeed, it's always the "mirror" animation of the "open".

File: src/win.c (hunk 3 above)
Comment (fdev31) [reply src/win.c:515]: This is on purpose, I couldn't decide, I left what I liked the most as the code but it's quite opinionated so I left the more boring but less debatable way in comments ;)

File: src/win.c (hunk 4 above)
Comment (fdev31) [reply src/win.c:518]: Yes, same purpose... I really like those animations but maybe it breaks on some toolkits / WM so I left the boring way in comments.

File: src/options.c (hunk 6 above)
Comment (fdev31) [reply src/options.c:190]: Ok, I switched to hyprland in the meantime and quite happy with it so I'm not sure I'll spend a lot of time on this PR, I can do some tiny things but maybe not more... note that most of the code isn't from me, I just simplified some of the code  and added a couple of things after fixing it for my own use case.

Looks like FTLabs made another branch which could be worth checking: https://github.com/FT-Labs/picom/tree/generalanimation

**Comment by r0-zero** [comment]
Hey guys,

I'll try to implement your ideas after all bug's are resolved

**Comment by colonelpanic8** [comment]
> Hey guys,
> 
> I'll try to implement your ideas after all bug's are resolved

It seems like the right approach should just be to get SOMETHING merged, and then iterate from there.

**Comment by colonelpanic8** [comment]
Also, can anyone weigh in on what the status of this PR is vs #772?

Are they just completely unrelated attempts to add animations?

Leaving outstanding PRs like this forever that don't get merged is bad for the community, IMO.

**Comment by dccsillag** [comment]
This PR is further work on top of what was done on #772. I haven't been able to check all of the differences between them, though.

**Comment by colonelpanic8** [comment]
@dccsillag weird are you sure? I don't see of the original commits from #772 in here.

**Comment by dccsillag** [comment]
I'm fairly sure that @FT-Labs's fork is built on top of mine (in fact, I think they were going to send a PR my way? I may be misremembering that though). It's quite likely that someone did an oopsie with git in the middle. @FT-Labs @fdev31 can you confirm?

**Comment by fdev31** [comment]
> It's quite likely that someone did an oopsie with git in the middle.

I can't confirm for sure, but I struggled to take over FT-lab's code and rebase it on Picom because it was mostly "hard to track" merges, so it's possible. And  I probably didn't help as well, with my own "mega patch"... so it's quite likely the history is messed up.

In short, my PR doesn't keep FTlab's history and focused on making code compatible with Picom's upstream, dealing with the history of FTlab wasn't easy (without making a huge fake merge) so I just mentioned it in github to keep track of it. Check FTlab's fork to know more about the history, you can assume my branch was 100% in sync with FTlab's when this PR was made (in case you need to check commit dates).

**Comment by colonelpanic8** [comment]
> I can't confirm for sure, but I struggled to take over FT-lab's code and rebase it on Picom because it was mostly "hard to track" merges, so it's possible. And I probably didn't help as well, with my own "mega patch"... so it's quite likely the history is messed up.

But I had already noted that. I looked at @FT-Labs 's code and I think it also does not include @dccsillag 's original commits either.

git was designed to work in a distributed fashion. Kind of frustrating that people are just taking code without having the original commits/attribution, and it also can land you in legal trouble eventually.

**Comment by colonelpanic8** [comment]
@FT-Labs is it worth it for me to attempt a merge of your current generalanimation branch?

**Comment by fdev31** [comment]
> Kind of frustrating that people are just taking code without having the original commits/attribution

I fully agree, but once the mess is done it's very hard to clean, especially when you are not the author... if I could have done a better job starting from this state please share some hints. It already took me quite some time with the indentation changes etc which were not kept in history so in the end I decided to trash the fork history.

**Pull request closed without merging**